### PR TITLE
MANTA-5146 schema-manager got stuck with no connections

### PR DESCRIPTION
**Pull request by bowrocker**
This changes fixes the re-try bug identified in MANTA-5146, where the cueball DNS resolver was giving up after only attempting to resolve a record from the first available resolver, causing schema manager to fail if the first nameservice zone was not available during provisioning.

Additionally, it introduces a refactoring of the entire resolver to extract the TrustDNS client code from the state machine and move it behind a trait implementation that hides the details of the client from the core code, so that if a different resolver were even desired it could be more easily swapped out.

This new trait also allows unit tests to implement mock revolvers more easily and removes external dependencies from them, which should both speed up testing as well as make them more reliable.

### CODE REVIEW

File: resolvers/dns-resolver/src/dns_client.rs [hunk 1]
```diff
@@ -0,0 +1,245 @@
...
+    }
+}
+
+pub fn init_dns_client(
+    resolver: &str,
+) -> Result<Arc<dyn DnsClient + Send + Sync>, ResolverError> {
+    let server = resolver.to_string().parse()?;
+    match UdpClientConnection::new(server) {
+        Ok(conn) => Ok(Arc::new(TrustDnsClient::from(SyncClient::new(conn)))),
+        Err(e) => Err(ResolverError::DnsClientError { err: e.to_string() }),
```
Comment (rjloura) [inline resolvers/dns-resolver/src/dns_client.rs:135]: isn't this covered by your `impl<ClientError> for ResolverError` above on line 117?

File: resolvers/dns-resolver/src/resolver.rs [hunk 2]
```diff
@@ -0,0 +1,545 @@
...
+    name: &str,
+    context: &mut ResolverContext,
+) -> Result<(), ResolverError> {
+    for resolver in context.resolvers.iter_mut() {
+        let srv_resp = match resolver.query_srv(&name, &context.log) {
+            Ok(r) => r,
+            Err(e) => {
+                error!(
+                    context.log,
+                    "failed to resolver srv: {}, trying next resolver",
```
Comment (isaacdavis) [inline resolvers/dns-resolver/src/resolver.rs:388]: Nit: should "resolver" be "resolve" here?

File: resolvers/dns-resolver/src/dns_client.rs (hunk 1 above)
Comment (bowrocker) [reply resolvers/dns-resolver/src/dns_client.rs]: Fixed, thanks.

File: resolvers/dns-resolver/src/resolver.rs (hunk 2 above)
Comment (bowrocker) [reply resolvers/dns-resolver/src/resolver.rs]: Fixed, good catch!

File: resolvers/dns-resolver/src/dns_client.rs [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+    let mut buf = Vec::new();
+    let mut f = File::open(DEFAULT_RESOLV_CONF)?;
+    f.read_exact(&mut buf)?;
+    let cfg = match resolv_conf::Config::parse(&buf) {
+        Ok(c) => c,
+        Err(e) => {
+            return Err(ResolverError::ResolvConfInvalid { err: e.to_string() })
+        }
+    };
+    for ns in cfg.nameservers.iter() {
```
[the remark this answers is not in the document]
Comment (bowrocker) [reply resolvers/dns-resolver/src/dns_client.rs]: Fixed, hope it's what you had in mind.

File: resolvers/dns-resolver/src/resolver.rs [hunk 4]
```diff
@@ -0,0 +1,551 @@
...
+        let srv_retry_params = RetryParams {
+            max: 0,
+            count: 0,
+            timeout: 1000,
+            min_delay: 0,
+            delay: 1000,
+            max_delay: 10000,
+        };
+
+        let srv_rec = SrvRec {
```
[the remark this answers is not in the document]
Comment (bowrocker) [reply resolvers/dns-resolver/src/resolver.rs]: Done.

File: resolvers/dns-resolver/src/resolver.rs [hunk 5]
```diff
@@ -0,0 +1,551 @@
...
+use tokio::prelude::*;
+
+use crate::dns_client::configure_default_resolvers;
+use crate::dns_client::configure_from_resolv_conf;
+use crate::dns_client::DnsClient;
+
+#[derive(Debug)]
+pub struct BackendRemovedMsg(pub backend::BackendKey);
+
+pub enum BackendAction {
```
[the remark this answers is not in the document]
Comment (bowrocker) [reply resolvers/dns-resolver/src/resolver.rs]: Done.

File: resolvers/dns-resolver/src/resolver.rs [hunk 6]
```diff
@@ -0,0 +1,551 @@
...
+use state_machine_future::{transition, RentToOwn, StateMachineFuture};
+use thiserror::Error;
+use tokio::prelude::*;
+
+use crate::dns_client::configure_default_resolvers;
+use crate::dns_client::configure_from_resolv_conf;
+use crate::dns_client::DnsClient;
+
+#[derive(Debug)]
+pub struct BackendRemovedMsg(pub backend::BackendKey);
```
[the remark this answers is not in the document]
Comment (bowrocker) [reply resolvers/dns-resolver/src/resolver.rs]: No reason, done.

File: resolvers/dns-resolver/src/resolver.rs [hunk 7]
```diff
@@ -37,6 +39,9 @@ impl DnsResolver {
         domain: String,
         service: String,
         resolvers: Option<Vec<Arc<dyn DnsClient + Send + Sync>>>,
+        /*
```
Comment (kellymclaughlin) [inline resolvers/dns-resolver/src/resolver.rs:42]: Remove?

**Pull request merged**